pyctbgui/utils/numpyWriter/npy_writer.py

```diff
@@ -0,0 +1,183 @@
+
+        self.bufferMax = bufferMax
+        self.dtype = np.dtype(dtype)  # in case we pass a type like np.float32
+        self.frameShape = frameShape
+        self.frameCount = 0
+        if not isinstance(file, zipfile.ZipExtFile):
+            fileExist = Path.is_file(Path(file))
+        else:
+            fileExist = True
+
+        newFile = resetFile or not fileExist
+        self.buffer = bytearray()
+        self.bufferCount = 0
+        # if newFile frameShape and dtype should be present
+        if newFile:
+            assert frameShape is not None
+            assert dtype is not None
+            # create/clear the file with mode wb+
+            self.file = open(file, 'wb+')
+            self.updateHeader()
+
+        else:
+            # opens file for read/write and check if the header of the file corresponds to the given function
+            # arguments
+            if not isinstance(file, zipfile.ZipExtFile):
+                self.file = open(file, 'rb+')
+            else:
+                self.file = file
+
+            self.file.seek(10)
+            headerStr = self.file.read(np.uint16(self.headerLength - 10)).decode("UTF-8")
+            header_dict = ast.literal_eval(headerStr)
+            self.frameShape = header_dict['shape'][1:]
+            if frameShape is not None:
+                assert frameShape == self.frameShape, \
+                    f"shape in arguments ({frameShape}) is not the same as the shape of the stored " \
+                    f"file ({self.frameShape})"
+
+            self.dtype = np.lib.format.descr_to_dtype(header_dict['descr'])
+            if dtype is not None:
+                assert dtype == self.dtype, \
+                    f"dtype in argument ({dtype}) is not the same as the dtype of the stored file ({self.dtype})"
+
+            self.frameCount = header_dict['shape'][0]
+
+            assert not header_dict['fortran_order'], "fortran_order in the stored file is not False"
+
+        self.__frameSize = np.dtype(self.dtype).itemsize * np.prod(self.frameShape)
+
+    def updateHeader(self):
+        """
+        updates the header of the .npy file with the class attributes
+        @note: fortran_order is always set to False
+        """
+        self.file.seek(0)
+        header_dict = {
+            'descr': np.lib.format.dtype_to_descr(self.dtype),
+            'fortran_order': False,
+            'shape': (self.frameCount, *self.frameShape)
+        }
+        np.lib.format.write_array_header_1_0(self.file, header_dict)
+
+    def writeOneFrame(self, frame: np.ndarray):
+        """
+        write one frame without buffering
+        @param frame: numpy array for a frame
+        """
+        assert frame.shape == self.frameShape
+        assert frame.dtype == self.dtype
+        self.file.seek(0, self.FSEEK_FILE_END)
+        self.frameCount += 1
+        self.file.write(frame.tobytes())
+        self.updateHeader()
+
+    def addFrame(self, frame: np.ndarray):
```

Comment:
The addition of a buffer makes the class more complicated did you do performance tests to see that we need it or how did you reason about this feature. I think for our use case writing directly only could be ok... but I can be convinced.